akka-actor/src/main/java/akka/japi/pf/ReceiveBuilder.java

```diff
@@ -32,78 +35,147 @@
  *
  * This is an EXPERIMENTAL feature and is subject to change until it has received more real world testing.
  */
-public class ReceiveBuilder {
+public class ReceiveBuilder extends AbstractPFBuilder<Object, BoxedUnit> {
+
   private ReceiveBuilder() {
   }
 
   /**
-   * Return a new {@link UnitPFBuilder} with no case statements. They can be added later as the returned {@link
-   * UnitPFBuilder} is a mutable object.
+   * Return a new {@link ReceiveBuilder} with no case statements. They can be
+   * added later as the returned {@link ReceiveBuilder} is a mutable object.
    *
    * @return a builder with no case statements
    */
-  public static UnitPFBuilder<Object> create() {
-    return new UnitPFBuilder<>();
+  public static ReceiveBuilder create() {
+    return new ReceiveBuilder();
   }
 
+  // FIXME
+  //  public static PartialFunction<Object, BoxedUnit> onMessage(FI.UnitApply<Object> apply) {
+  //    return matchAny(apply).build();
+  //  }
+
   /**
-   * Return a new {@link UnitPFBuilder} with a case statement added.
+   * Add a new case statement to this builder.
    *
-   * @param type  a type to match the argument against
-   * @param apply an action to apply to the argument if the type matches
+   * @param type
+   *          a type to match the argument against
+   * @param apply
+   *          an action to apply to the argument if the type matches
    * @return a builder with the case statement added
    */
-  public static <P> UnitPFBuilder<Object> match(final Class<? extends P> type, FI.UnitApply<? extends P> apply) {
-    return UnitMatch.match(type, apply);
+  @SuppressWarnings("unchecked")
+  public <P> ReceiveBuilder match(final Class<? extends P> type, final FI.UnitApply<? extends P> apply) {
```

Comment:
Are we really OK with `type` and `apply` not actually needing to be the same type? It allows for subtle compile errors. If we do I'd at least prefer a warning in the docs. To wrap up my earlier research:
- `P` is not bound to anything, so java is free to pick `Object`
- The `? extends P` was introduced to allow matching on the raw class of a generic message, e.g. `.match(List.class, (List<String> msg) -> ...)`
- However, this actually disconnects the two types completely, so one could have 
```
void handle(Msg1 m) { ... }
void handle(Msg2 m) { ... }
receiveBuilder.match(Msg1.class, this::handle)
```
and it wouldn't compile, since both `Msg1` and `Msg2` match `? extends Object`.

Similarly, you can do `.match(String.class, (Integer i) -> ...)` which would compile fine, but throw `ClassCastException` at runtime.

I've suggested the following before, I think after discussion with @ktoso:
1. Rename the existing method to `.matchUnchecked`, and simply declare both types `<?>` instead of `<? extends P>`
2. Have a new, actually type-safe method `<P> ReceiveBuilder match (final Class<P> type, final FI.UnitApply<P> apply)`

This comment also applies to `match` on `? extends P` with predicate.